### PR TITLE
[SPARK-45649][SQL] Unify the prepare framework for `OffsetWindowFunctionFrame`

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Currently, the implementation of the `prepare` of all the `OffsetWindowFunctionFrame` have the same code logic show below.
```
  override def prepare(rows: ExternalAppendOnlyUnsafeRowArray): Unit = {
    if (offset > rows.length) {
      fillDefaultValue(EmptyRow)
    } else {
      resetStates(rows)
      if (ignoreNulls) {
        ...
      } else {
        ...
      }
    }
  }
```
This PR want unify the prepare framework for `OffsetWindowFunctionFrame`

### Why are the changes needed?
Unify the prepare framework for `OffsetWindowFunctionFrame`

### Does this PR introduce _any_ user-facing change?
'No'.
Inner update.

### How was this patch tested?
Exists test cases.

### Was this patch authored or co-authored using generative AI tooling?
'No'.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala [hunk 1]
```diff
@@ -175,6 +176,23 @@ abstract class OffsetWindowFunctionFrameBase(
     }
   }
 
+  override def prepare(rows: ExternalAppendOnlyUnsafeRowArray): Unit = {
+    if (offset > rows.length) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala:180]: shall we use `Math.abs(offset)`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala [hunk 2]
```diff
@@ -284,7 +293,7 @@ class FrameLessOffsetWindowFunctionFrame(
       }
   } else {
     (current: InternalRow) =>
-      if (inputIndex >= 0 && inputIndex < input.length) {
+      if (inputIndex >= 0 && input != null && inputIndex < input.length) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala:296]: why this change?

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala:296]: If `Math.abs(offset) > rows.length` is true, no need to assign a value to `input`.

**Comment by beliefer** [comment]
The GA failure is unrelated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala:297]: Why we don't need the null check before? where did we instantiate `input`?

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala:297]: Because the origin code always call `resetStates`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala [hunk 3]
```diff
@@ -175,6 +178,23 @@ abstract class OffsetWindowFunctionFrameBase(
     }
   }
 
+  override def prepare(rows: ExternalAppendOnlyUnsafeRowArray): Unit = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala:181]: shall we overwrite `doWrite` do skip if `input == null`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala:297]: I feel the code is a bit subtle now. In the branch above here, there is no null check: `while (nextSelectedRow == EmptyRow && inputIndex < input.length) {`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala (hunk 3 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala:181]: Got it.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@cloud-fan Thank you!

**Comment by cloud-fan** [comment]
This has been reverted. See the reason in https://issues.apache.org/jira/browse/SPARK-45649

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala [hunk 4]
```diff
@@ -196,24 +224,15 @@ class FrameLessOffsetWindowFunctionFrame(
...
     ignoreNulls: Boolean = false)
   extends OffsetWindowFunctionFrameBase(
-    target, ordinal, expressions, inputSchema, newMutableProjection, offset) {
+    target, ordinal, expressions, inputSchema, newMutableProjection, offset, ignoreNulls) {
 
-  override def prepare(rows: ExternalAppendOnlyUnsafeRowArray): Unit = {
-    resetStates(rows)
-    if (ignoreNulls) {
-      if (Math.abs(offset) > rows.length) {
-        fillDefaultValue(EmptyRow)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala:205]: This PR is not a pure refactor. Some logic is changed. e.g. before this PR, `fillDefaultValue(EmptyRow)` is never called if `ignoreNulls == false`. We need to understand the code better before making these changes.
